src/include/rados/librados.h

```diff
@@ -1481,6 +1481,24 @@ CEPH_RADOS_API uint64_t rados_get_last_version(rados_ioctx_t io);
 CEPH_RADOS_API int rados_write(rados_ioctx_t io, const char *oid,
                                const char *buf, size_t len, uint64_t off);
 
+/**
+ * Write *len* bytes from *buf* into the *oid* object, starting at
+ * offset *off*. The value of *len* must be <= UINT_MAX/2.
+ *
+ * This function is not thread-safe, i.e. contents of buf must remain
+ * the same during entire function call.
+ *
+ * @note This will never return a positive value not equal to len.
+ * @param io the io context in which the write will occur
+ * @param oid name of the object
+ * @param buf data to write
+ * @param len length of the data, in bytes
+ * @param off byte offset in the object to begin writing at
+ * @returns 0 on success, negative error code on failure
+ */
+CEPH_RADOS_API int rados_write2(rados_ioctx_t io, const char *oid,
```

Comment:
Nit: I question whether or not new methods are required. In theory it should be safe to assume that the contents of ```buf``` for an AIO method must remain the same until the call completes (similar to ```aio_write``` in Linux [1]). Plus, removing the const-ness of the ```buf``` seems odd.

[1] http://man7.org/linux/man-pages/man3/aio_write.3.html

Reply:
@dillaman it turns out the input buffer is indeed *deep* copied by librados. see `buffer::list::append(const char *data, unsigned len)`.

and i am not sure we can make this semantic change even we bump up the soversion.

Reply:
Yeah, the original `rados_write` functions deep copy the `buf` because of `buffer::list::append`, I agree with dilliaman that it is safe to assume the buf passed in `rados_write`, `rados_aio_write`, etc should be valid before io finishes, just like `write` and `aio_write` in Linux, thus I suggest updating the original `rados_write` by replacing `buffer::list::append` with `buffer::create_static`.

Reply:
@liyichao if we are sure that all the librados clients can ensure the the buffer passed to these write calls will be valid until the call completes. there is no reason we cannot do the switch. 

but to be 100% sure about this, i'd suggest check the callers of librados, like librbd,libcephfs,python-rados and qemu.